Implement FEATURE `suppress-bindist-buildpkg`

Kangie opened the pull request:
This feature combines with `buildpkg` to enable the installation of packages with "RESTRICT=bindist" without the creation of a binpkg.

This should help binhost builders improve the library of pre-packaged software without running the risk of accidentally (bin)packaging a dependency that we can't legally distribute (or don't want to pay royalties on!)

**Current Behaviour**:
FEATURES=buildpkg + RESTRICT=bindist
	=> Package is installed, binpkg created
FEATURES=buildpkg + RESTRICT=bindist + ACCEPT_RESTRICT=-bindist
	=> Package is masked, cannot be selected

**New Behaviour**:
FEATURES=buildpkg + RESTRICT=bindist
	=> Package is installed, binpkg created
FEATURES=buildpkg suppress-bindist-buildpkg + RESTRICT=bindist
	=> Package is installed, binpkg creation stage skipped
FEATURES=* + RESTRICT=bindist + ACCEPT_RESTRICT=-bindist
	=> Any combination of features and the above are masked

Bug: https://bugs.gentoo.org/542480

Kangie committed:
Implement FEATURE `suppress-bindist-buildpkg`

This feature combines with `buildpkg` to enable the installation of
packages with "RESTRICT=bindist" without the creation of a binpkg.

This should help binhost builders improve the library of pre-packaged
software without running the risk of accidentally (bin)packaging a dependency
that we can't legally distribute (or don't want to pay royalties on!)

Old Behaviour:
FEATURES=buildpkg + RESTRICT=bindist
	=> Package is installed, binpkg created
FEATURES=buildpkg + RESTRICT=bindist + ACCEPT_RESTRICT=-bindist
	=> Package is masked, cannot be selected

New Behaviour:
FEATURES=buildpkg + RESTRICT=bindist
	=> Package is installed, binpkg created
FEATURES=buildpkg suppress-bindist-buildpkg + RESTRICT=bindist
	=> Package is installed, binpkg creation stage skipped
FEATURES=* + RESTRICT=bindist + ACCEPT_RESTRICT=-bindist
	=> Any combination of features and the above are masked

Bug: https://bugs.gentoo.org/542480
Signed-off-by: Matt Jolly <kangie@gentoo.org>

thesamesam commented on man/make.conf.5:
```diff
@@ -796,6 +796,14 @@ Have portage react strongly to conditions that may conflict with system
 security provisions (for example textrels, executable stack).  Read about
 the \fIQA_STRICT_*\fR variables in \fBmake.conf\fR(5).
 .TP
+.B suppress\-bindist\-binpkg
```
Maybe add something about how it only does something if ACCEPT_RESTRICT excludes bindist? I know that's obvious, but it makes things clearer for users trying to understand how things fit together.

thesamesam commented on lib/portage/tests/resolver/test_accept_restrict_bindist.py:
```diff
@@ -0,0 +1,300 @@
+            },
+        }
+
+        # Test default behavior (ACCEPT_RESTRICT="*" - should accept all)
```
Thanks for the nice comments, I like seeing these for testcases especially.

thesamesam commented on lib/portage/tests/emerge/test_buildpkg.py:
```diff
@@ -0,0 +1,483 @@
+            success = self._run_commands(playground, test_commands, debug)
+            self.assertTrue(
+                success,
+                "Package with RESTRICT=bindist should be installable with suppress-bindist-buildpkg feature",
```
... where we don't have ACCEPT_RESTRICT="... -bindist", right?

thesamesam commented on man/make.conf.5:
```diff
@@ -796,6 +796,14 @@ Have portage react strongly to conditions that may conflict with system
+.B suppress\-bindist\-binpkg
+When enabled, this feature prevents the creation of binary packages for
+ebuilds that have the \fIbindist\fR restriction. This allows users to
+install such packages normally while ensuring that no binary packages
```
Perhaps explicitly state what happens when it's disabled (the default)?

Kangie committed:
Increase minimum python version to 3.10

3.9 has not been a valid python target in ::gentoo for a few years,
it's probably safe to start using python 3.10 features.

Signed-off-by: Matt Jolly <kangie@gentoo.org>

Kangie commented on pyproject.toml:
```diff
@@ -3,7 +3,7 @@ name = 'portage'
 description = 'Portage is the package management and distribution system for Gentoo'
 readme = 'README.md'
-requires-python = '>=3.9'
+requires-python = '>=3.10'
```
Well, since you got here before I had a chance to comment:

I've bumped the minimum requirement to py3.10 as this change fails with py3.9 (I used `match` in a test), which wasn't picked up until CI runs because I don't develop against EoL'd (or almost EoL'd) Python.

It wouldn't be hard to rework the code to work with older python, but I also don't see much point in making such a change if we're not actually sure that we want py3.9 anyway. Since we're reviewing changes anyway, I opted to drop the older version (which was privately discussed earlier as a possibility). I'm equally happy to just use `if/else` if we _really_ want old python support.
